File: src/hello_world/calculator.py
# import math # unused import for testing ruff


class Calculator:
    def __init__(self) -> None:
        self.reset()

    def add(self, value: float) -> float:
        self.total += value
        return self.getValue()

    def sub(self, value: float) -> float:
        self.total -= value
        return self.getValue()

    def mul(self, value: float) -> float:
        self.total *= value
        return self.getValue()

    def div(self, value: float) -> float:
        if value == 0:
            raise ValueError("Cannot divide by zero")
        self.total /= value
        return self.getValue()

    def reset(self) -> None:
        self.total: float = 0

    def getValue(self) -> float:
        return self.total

File: src/cpp/calculator.cpp
#include "calculator.h"

#include <stdexcept>

namespace hello_world {

Calculator::Calculator() : total(0.0f) {}

const float Calculator::add(const float value) {
    total += value;
    return total;
}

const float Calculator::sub(const float value) {
    total -= value;
    return total;
}

const float Calculator::mul(const float value) {
    total *= value;
    return total;
}

const float Calculator::div(const float value) {
    if (value != 0.0f) {
        total /= value;
    } else {
        throw std::invalid_argument("Cannot divide by zero");
    }
    return total;
}

void Calculator::reset() {
    total = 0.0f;
}

const float Calculator::getValue() const {
    return total;
}

} // namespace hello_world
